text/unification.md

```diff
@@ -0,0 +1,268 @@
+where:
+- `a₁:A₁,...,aₙ:Aₙ` is the defining context of the existential variable (called `evar_hyps` in the implementation, it is made of named declarations, possibly with local definitions too, so, shortly a `named_context`)
+- `B` is its type (`evar_concl`)
+
+To be used in a term, existential variables need to be applied to a substitution of its context. So, they take the form `?e@{a₁:=t₁,...,aₙ:tₙ}` where `a₁:=t₁,...,aₙ:=tₙ` is a *substitution* (sometimes called *instance*, often written `args` in the code).
+
+## Existential variables map
+
+A collection of declarations of existential variables is conventionally written `Σ` (an `evar_map` in the code, using either `sigma` or `evd` to refers to it).
+
+## Unification problems
+
+Given an existential variable map, *unification problems* are equations of the form `Γ ⊢ t₁ =?= t₂` or inequations of the form `Γ ⊢ t₁ ≤? t₂` where `t₁` and `t₂` are terms possibly mentioning existential variables with their substitution and both terms living in a context of variables `Γ` (that is the juxtaposition of a `named_context` and of a `rel_context`). An inequation corresponds to a subtyping problem between types (e.g. `Prop ≤ Type`). Note that we assume the existence of a global context of inductive types, definitions and axioms common to the definitions of existential variables and to the unification problems.
+
+A unification algorithm turns a `Σ` and a collection of unification problems into either an impossibility of satisfaction of the equations, or in a collection of (well-typed) definitions for the existential variables such that all problems are satisfied, or a residual set of unification problems that the algorithm is neither able to invalidate nor to satisfy (since unification in general is undecidable in the Calculus of Inductive Constructions).
+
+## Reflexive unification problems
+
+We call *reflexive unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= ?e@{subst'}`.
+
+## Atomic unification problems
+
+We call *atomic unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= ?e'@{subst'}` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? ?e'@{subst'}`.
+
+## Simple unification problems
+
+We call *simple unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= t` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? t`. Unification basically reduces arbitrary complex unification problems to simple ones which in turn may generate new arbitrary complex problems.
+
+## Simplification of reflexive unification problems
+
+A reflexive unification problem induces constraints on the possible definitions of `?e`: once `?e` is defined, it may generate new unification problems between terms of the instances that contain existential variables.
+
+A reflexive unification problem can be simplified as follows:
+- Variables of the context whose instances in the substitution are provably incompatible (e.g. because corresponding to distinct head normal forms) can be removed from the context.
```

Comment:
This optimization seems unsound?
Consider `x : nat |- ?e : nat` and the pair of unification problems `|- ?e@{x:=O} == ?e@{x:=S O}` and `x : nat |- ?e@{x:=x} == match x with O => O | S _ => O end`.  The only valid solution of this is `?e := match x with O => O | S _ => O end`, which refers to `x`.
And in fact Coq fails to solve such a problem:
```coq
Check let e := (fun x => ?[e]) in
      eq_refl : (e O, e) = (e (S O), fun x => match x with O => O | S _ => O end).
(* Error:
In environment
e := fun _ : nat => ?e : nat -> nat
The term "eq_refl" has type "(e 0, e) = (e 0, e)" while it is expected to have type
 "(e 0, e) = (e 1, fun x : nat => match x with
                                  | 0 | _ => 0
                                  end)" (cannot instantiate "?e" because "x" is not in its scope).
*)
```

Reply:
Reported as https://github.com/coq/coq/issues/15110